### ISSUE TITLE
Make the model pass validation reliably (ID vault + SQL extraction + sentinel + loss masking + tokenizer specials)

### ISSUE THREAD

**Issue by AndrewMichael2020** [AndrewMichael2020/my-gpt-2-2#7]
While test-running colab_train.ipynb, we hit the block:

`Loaded checkpoint from epoch 3

================================================================================
GENERATING SQL FOR TEST QUESTIONS
================================================================================

[Question 1]
Original: How many visits did patient 5432 have in department 25?
Clean: How many visits did patient __ID_1__ have in department 25?
ID Map: {'__ID_1__': '5432'}

Generated SQL (with placeholders):
  SCHEMA : Patients ( PatientID , FirstName , LastName , DateOfBirth , Gender , InsuranceProvider ); Visits ( VisitID , PatientID , VisitDate , DepartmentID , ProviderID , VisitType , TotalCharge ); Departments ( DepartmentID , DepartmentName , Location ); Providers ( ProviderID , ProviderName , Specialty , DepartmentID ); Diagnoses ( DiagnosisID , VisitID , ICDCode , DiagnosisDescription ) QUESTION : How many visits did patient __ID_1__ have in department 25 ? ID_MAP : __ID_1__ = 54 32 SQL : SELECT COUNT (*) FROM Visits WHERE PatientID = __ID_1__ AND DepartmentID = __ID_2__ ;
✗ Validation: FAILED
  - Must contain exactly one statement ending with ';'
  - Must start with SELECT
  - Unknown placeholder: __ID_2__
--------------------------------------------------------------------------------

[Question 2]
Original: What is the total charge for patient 7890?
Clean: What is the total charge for patient __ID_1__?
ID Map: {'__ID_1__': '7890'}

Generated SQL (with placeholders):
  SCHEMA : Patients ( PatientID , FirstName , LastName , DateOfBirth , Gender , InsuranceProvider ); Visits ( VisitID , PatientID , VisitDate , DepartmentID , ProviderID , VisitType , TotalCharge ); Departments ( DepartmentID , DepartmentName , Location ); Providers ( ProviderID , ProviderName , Specialty , DepartmentID ); Diagnoses ( DiagnosisID , VisitID , ICDCode , DiagnosisDescription ) QUESTION : What is the total charge for patient __ID_1__ ? ID_MAP : __ID_1__ = 78 90 SQL : SELECT SUM ( TotalCharge ) FROM Visits WHERE PatientID = __ID_1__ ;
✗ Validation: FAILED
  - Must contain exactly one statement ending with ';'
  - Must start with SELECT
--------------------------------------------------------------------------------

[Question 3]
Original: Show monthly visit counts in 2023
Clean: Show monthly visit counts in __ID_1__
ID Map: {'__ID_1__': '2023'}

Generated SQL (with placeholders):
  SCHEMA : Patients ( PatientID , FirstName , LastName , DateOfBirth , Gender , InsuranceProvider ); Visits ( VisitID , PatientID , VisitDate , DepartmentID , ProviderID , VisitType , TotalCharge ); Departments ( DepartmentID , DepartmentName , Location ); Providers ( ProviderID , ProviderName , Specialty , DepartmentID ); Diagnoses ( DiagnosisID , VisitID , ICDCode , DiagnosisDescription ) QUESTION : Show monthly visit counts in __ID_1__ ID_MAP : __ID_1__ = 2023 SQL : SELECT MONTH ( VisitDate ) AS Month , COUNT (*) FROM Visits WHERE YEAR ( VisitDate ) = __ID_1__ GROUP BY MONTH ( VisitDate );
✗ Validation: FAILED
  - Must contain exactly one statement ending with ';'
  - Must start with SELECT
--------------------------------------------------------------------------------

================================================================================
RESULTS: 0/3 questions passed validation
================================================================================

✗ INCOMPLETE: Need more training or better templates
`

Now:

## Context
Current inference output frequently echoes `SCHEMA/QUESTION/ID_MAP` before the SQL, invents missing placeholders (e.g., `__ID_2__`), and sometimes omits the terminal semicolon. This causes the validator to fail with:
- “Must start with SELECT”
- “Unknown placeholder”
- “Must contain exactly one statement ending with ';'”

We will fix this aggressively with deterministic preprocessing/postprocessing plus training adjustments that teach **prompt → SQL**.

## Goal
Achieve **>= 90% validation pass rate** on a held-out test set of 100 questions using the current toy schema (Patients/Visits/Departments/Providers/Diagnoses), while preserving the constraints:
- output is exactly one statement
- SELECT-only
- ends with `;`
- placeholder integrity is perfect
- schema-known entities only

## Deliverables
1. Updated **ID vault** that extracts all relevant identifiers deterministically (including DepartmentID).
2. Robust **SQL extraction** that returns only the SQL statement even if the model echoes prompt sections.
3. Add an **end-of-SQL sentinel** token (e.g., `</SQL>`) and use it in dataset + inference stop condition.
4. Implement **loss masking** so training loss is computed only on tokens after `SQL:`.
5. Train tokenizer with required **special tokens** so placeholders and section markers are stable.
6. Add tests and an evaluation script to measure pass rate and common failure causes.

## Files and modules (target)
- `sql_agent/id_vault.py`
- `sql_agent/prompt.py`
- `sql_agent/generate.py`
- `sql_agent/validate.py`
- `sql_agent/train/dataset.py`
- `sql_agent/train/train_loop.py`
- `scripts/build_dataset.py`
- `scripts/train_tokenizer.py`
- `scripts/eval.py`
- `notebooks/colab_train.ipynb` (wire end-to-end)

### 1) ID vault: extract **all** identifiers deterministically
**Problem:** We replaced patient ID with `__ID_1__` but left `department 25` raw, so the model invented `__ID_2__`.

**Implement:**
- Update extraction to collect:
  - All integer tokens likely to be identifiers (recommend: 1–6 digits).
  - Code-like tokens (alphanumerics with `_` or `-`), optional.
  - ISO dates `YYYY-MM-DD`.
- Add year handling:
  - If integer is within `[1900..2100]` AND context includes “in”, “year”, “during”, treat as `__YEAR_1__`.
  - Otherwise treat as `__ID_n__`.
- Ensure stable ordering:
  - Placeholders assigned in order of first appearance in the question text.

**Acceptance checks:**
- Q: “How many visits did patient 5432 have in department 25?”
  - Clean: “… patient __ID_1__ … department __ID_2__ …”
  - Map includes both `5432` and `25`.

**Add tests:**
- `test_id_vault_extracts_multiple_ids()`
- `test_id_vault_year_detection_2023()`
- `test_id_vault_preserves_order()`

---

### 2) SQL extraction: return **only** the SQL statement from completion
**Problem:** Model echoes `SCHEMA/QUESTION/ID_MAP` so validator sees text that does not start with `SELECT`.

**Implement extraction function** (in `sql_agent/generate.py` or `sql_agent/validate.py`):
- Take model completion text.
- Find the first `SELECT` (case-insensitive) position.
- From there, take substring up to:
  - the first `;` after SELECT, OR
  - the sentinel `</SQL>` (after Step 3), OR
  - max length cap (fallback).
- Strip whitespace.
- Ensure the extracted SQL ends with `;` (if sentinel path is used, enforce `;` before `</SQL>` during dataset creation).

**Add tests:**
- `test_extract_sql_from_prompt_echo()`
- `test_extract_sql_handles_multiple_semicolons_reject()` (if extra `;` occurs, reject)

---

### 3) Add an explicit end-of-SQL sentinel: `</SQL>`
**Problem:** Some outputs omit `;` or continue producing extra text.

**Implement:**
- Choose sentinel token: `</SQL>`.
- Update dataset generation so each training target ends as:
  - `SQL: SELECT ... ; </SQL>`
- Update inference to stop generation when `</SQL>` appears.
- Update extraction to prefer sentinel-based truncation.

**Acceptance checks:**
- All generated training samples end with `;</SQL>` sequence (or `; </SQL>` depending on spacing).
- Inference stops early consistently and returns a single statement.

---

### 4) Loss masking: compute loss only on tokens after `SQL:`
**Problem:** Training on full concatenated record encourages the model to reproduce headers (`SCHEMA`, `QUESTION`, `ID_MAP`).

**Implement:**
- In dataset collation, create labels where tokens before `SQL:` are masked out:
  - set label to `-100` for prompt tokens
  - keep labels for SQL tokens including `</SQL>`
- Keep the input still containing schema/question/id_map, but only supervise SQL.

**Add tests:**
- `test_loss_masking_ignores_prompt_tokens()`
- Confirm the first unmasked label token aligns immediately after `SQL:`

---

### 5) Tokenizer: add required special tokens
**Problem:** Placeholders and IDs can be split oddly (“54 32”). Special tokens make patterns stable.

**Implement in `scripts/train_tokenizer.py`:**
- Add these as special tokens before training or ensure they are never split:
  - Section markers: `SCHEMA:`, `QUESTION:`, `ID_MAP:`, `SQL:`
  - Sentinel: `</SQL>`
  - Placeholder prefixes or full forms:
    - `__ID_`, `__DATE_`, `__YEAR_` (minimum)
  - Optionally reserve a range:
    - `__ID_1__` ... `__ID_64__`
    - `__DATE_1__` ... `__DATE_16__`
    - `__YEAR_1__` ... `__YEAR_8__`
- Ensure tokenizer save/load is consistent across training and inference.

**Add tests:**
- `test_tokenizer_keeps_sentinel_single_token()`
- `test_tokenizer_keeps_placeholder_prefix_intact()`

---

### 6) Update validator to match new pipeline
**Implement in `sql_agent/validate.py`:**
- Validate extracted SQL only (post-extraction).
- Enforce:
  - exactly one statement
  - starts with `SELECT` (after trimming)
  - ends with `;`
  - rejects DML/DDL keywords: INSERT/UPDATE/DELETE/MERGE/DROP/ALTER/CREATE/EXEC/GRANT/REVOKE/TRUNCATE
  - placeholder integrity:
    - any `__ID_\d+__`, `__DATE_\d+__`, `__YEAR_\d+__` must exist in ID vault
    - no unknown placeholders
- Keep schema-known table/column checks (string-level MVP is fine).

**Add tests:**
- `test_validator_rejects_unknown_placeholder()`
- `test_validator_accepts_clean_select_only()`
- `test_validator_rejects_dml_keywords()`

---

### 7) Evaluation harness: measure pass rate and failure reasons
**Implement `scripts/eval.py`:**
- Run N held-out questions through the pipeline.
- Report:
  - pass rate
  - top failure categories (start token, missing semicolon, unknown placeholder, schema mismatch)
- Save a small JSON report with examples of failures.

**Acceptance checks:**
- After a short training run (resume from checkpoint is fine), test pass rate >= 90% on 100 held-out questions.

## Definition of Done
- ID vault replaces **all** identifiers required by SQL (patient ID, department ID, provider ID, etc.).
- Model outputs are reliably post-processed to a single `SELECT ...;` statement.
- Sentinel-based stopping works.
- Loss masking prevents header echoing from dominating learning.
- Tokenizer preserves sentinel and placeholder patterns.
- Eval shows >= 90% validation pass rate on 100 held-out questions.

## Notes
- Keep the SQL feature set intentionally small.
- Prefer deterministic plumbing over “more training” until the pipeline is correct.
